src/linked_hash_map.rs

```diff
@@ -1677,6 +1749,173 @@ impl<'a, K, V> Drop for Drain<'a, K, V> {
     }
 }
 
+impl<'a, K, V, S> CursorMut<'a, K, V, S> {
+    /// Returns an `Option` of the current element in the list, provided it is not the
+    /// _guard_ node, and `None` overwise.
+    #[inline]
+    pub fn current(&mut self) -> Option<(&K, &mut V)> {
+        unsafe {
+            let at = NonNull::new_unchecked(self.cur);
+            self.peek(|| at)
+        }
+    }
+
+    /// Retrieves the next element in the list (moving towards the end).
+    #[inline]
+    pub fn peek_next(&mut self) -> Option<(&K, &mut V)> {
+        unsafe {
+            let at = (*self.cur).links.value.next;
+            self.peek(|| at)
+        }
+    }
+
+    /// Retrieves the previous element in the list (moving towards the front).
+    #[inline]
+    pub fn peek_prev(&mut self) -> Option<(&K, &mut V)> {
+        unsafe {
+            let at = (*self.cur).links.value.prev;
+            self.peek(|| at)
+        }
+    }
+
+    // Retrieves the element specified by the at closure function without advancing current
+    // position to it.
+    #[inline]
+    fn peek(&mut self, at: impl FnOnce() -> NonNull<Node<K, V>>) -> Option<(&K, &mut V)> {
+        if let Some(values) = self.values {
+            unsafe {
+                let node = at().as_ptr();
+                if node == values.as_ptr() {
+                    None
+                } else {
+                    let entry = (*node).entry_mut();
+                    Some((&entry.0, &mut entry.1))
+                }
+            }
+        } else {
+            None
+        }
+    }
+
+    /// Updates the pointer to the current element to the next element in the
+    /// list (that is, moving towards the end).
+    #[inline]
+    pub fn move_next(&mut self) {
+        let at = unsafe { (*self.cur).links.value.next };
+        self.muv(|| at);
+    }
+
+    /// Updates the pointer to the current element to the previous element in the
+    /// list (that is, moving towards the front).
+    #[inline]
+    pub fn move_prev(&mut self) {
+        let at = unsafe { (*self.cur).links.value.prev };
+        self.muv(|| at);
+    }
+
+    /// Positions the cursor at the element associated with the specified key. Returns `true`
+    /// if the element exists and the operation succeeds, or `false` if the element does not
+    /// exist.
+    #[inline]
+    pub fn move_at(&mut self, key: K) -> bool
```

Comment:
```suggestion
    pub fn move_at(&mut self, key: K) -> Option<(&K, &mut V)>
```

This API offers a more user-friendly experience as it provides the key-value in one go. If the caller isn't interested in inspecting or modifying it, they can simply ignore it. Plus, it clearly indicates whether the method call was successful.

What do you think?

Reply:
Regarding the use of `(&K, &mut V)`, I've observed multiple instances of `&mut K` being returned by functions. Could this potentially lead to synchronization issues, if the actual key in the node differs from the key used for its insertion into the HashTable?

In my implementation, I've ensured that the key cannot be modified once it's inserted. However, I might be overlooking some aspects. I'd appreciate your assistance in understanding this better.

Reply:
> Regarding the use of (&K, &mut V), I've observed multiple instances of &mut K being returned by functions. Could this potentially lead to synchronization issues, if the actual key in the node differs from the key used for its insertion into the HashTable?

These mostly mimic methods that are available in hashbrown, [like this one](https://docs.rs/hashbrown/latest/hashbrown/hash_map/struct.RawVacantEntryMut.html#method.insert) if the behavior of those methods differs from what hashbrown does, it's probably wrong.

I might be mistaken, but are you asking about the `&mut K`s being returned from methods in `RawVacantEntryMut`? Those actually behave the same as hashbrown afaik, but their behavior is not documented that well (or I just can't find it right now). When you receive a `RawVacantEntryMut`, you receive it for some key that *was not found*, and it's up to the user to use that `RawVacantEntryMut` to insert an entry with *the same key*. If you use `RawVacantEntryMut` to insert some *other* key, it will **mostly** work, except for the fact that you can use this to end up with duplicate entries in the table. The `&mut K` returned from `RawVacantEntryMut` methods always refers to a *newly* inserted entry, because it *always* inserts a new entry (even if this would result in duplicate entries).

I actually missed this entirely when reviewing #21, I'm ashamed to say, but it is afaict still acceptable behavior for the *raw* table API.